## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

In [2]:
heart_df = heart_df=pd.read_csv("C:\\Users\\Ankita & Subhash\\Desktop\\MSITM\\Fall Semester 2018\\Machine Learning\\Project\\Data From Competition Website\\train_values.csv")
heart_df.head()

patient_id  slope_of_peak_exercise_st_segment               thal  \
0     0z64un                                  1             normal   
1     ryoo3j                                  2             normal   
2     yt1s1x                                  1             normal   
3     l2xjde                                  1  reversible_defect   
4     oyt4ek                                  3  reversible_defect   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     128                2                  0   
1                     110                3                  0   
2                     125                4                  3   
3                     152                4                  0   
4                     178                1                  0   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          308                       0.0    1   45   
1                          214                       1.6    0   54   
2                          304                       0.0    1   77   
3                          223                       0.0    1   40   
4                          270                       4.2    1   59   

   max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  
0                      170                        0                      0  
1                      158                        0                      0  
2                      162                        1                      1  
3                      181                        0                      1  
4                      145                        0                      0

In [10]:
## Convert Categorical Variable to Numeric Variable
heart_df["thal_cleaned"]=np.where(heart_df["thal"]=="normal",0,
                                  np.where(heart_df["thal"]=="fixed_defect",1,
                                           np.where(heart_df["thal"]=="reversible_defect",2,3)
                                          )
                                 )

In [9]:
heart_df.head(10)

patient_id  slope_of_peak_exercise_st_segment               thal  \
0     0z64un                                  1             normal   
1     ryoo3j                                  2             normal   
2     yt1s1x                                  1             normal   
3     l2xjde                                  1  reversible_defect   
4     oyt4ek                                  3  reversible_defect   
5     ldukkw                                  1             normal   
6     2gbyh9                                  2  reversible_defect   
7     daa9kp                                  2       fixed_defect   
8     3nwy2n                                  3  reversible_defect   
9     1r508r                                  2             normal   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     128                2                  0   
1                     110                3                  0   
2                     125                4                  3   
3                     152                4                  0   
4                     178                1                  0   
5                     130                3                  0   
6                     150                4                  2   
7                     150                4                  1   
8                     170                4                  0   
9                     120                3                  0   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   
5                                     0                    0   
6                                     0                    2   
7                                     0                    2   
8                                     0                    2   
9                                     0                    0   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          308                       0.0    1   45   
1                          214                       1.6    0   54   
2                          304                       0.0    1   77   
3                          223                       0.0    1   40   
4                          270                       4.2    1   59   
5                          180                       0.0    1   42   
6                          258                       2.6    0   60   
7                          276                       0.6    1   57   
8                          326                       3.4    1   59   
9                          219                       1.6    0   50   

   max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  \
0                      170                        0                      0   
1                      158                        0                      0   
2                      162                        1                      1   
3                      181                        0                      1   
4                      145                        0                      0   
5                      150                        0                      0   
6                      157                        0                      1   
7                      112                        1                      1   
8                      140                        1                      1   
9                      158                        0                      0   

   thal_cleaned  
0             0  
1             0  
2             0  
3             2  
4             2  
5             0  
6             2  
7             1  
8             2  


Okay, so we got our Categorical Variable **thal** converted into numerical variable, **thal_cleaned** for the sake of fitting the Gaussian Naive Bayes.

In [17]:
# Split dataset in training and test datasets
X_train, X_test = train_test_split(heart_df, test_size=0.2, random_state=int(time.time()))

In [18]:
# Instantiate the classifier
gnb = GaussianNB()
used_features =[
    "slope_of_peak_exercise_st_segment",
    "thal_cleaned",
    "resting_blood_pressure",
    "chest_pain_type",
    "num_major_vessels",
    "fasting_blood_sugar_gt_120_mg_per_dl",
    "resting_ekg_results",
    "serum_cholesterol_mg_per_dl",
    "oldpeak_eq_st_depression",
    "sex",
    "age",
    "max_heart_rate_achieved",
    "exercise_induced_angina"
]

# Train classifier
gnb.fit(
    X_train[used_features].values,
    X_train["heart_disease_present"]
)
y_pred = gnb.predict(X_test[used_features])

# Print results
print("Number of mislabeled points out of a total {} points : {}, performance {:05.2f}%"
      .format(
          X_test.shape[0],
          (X_test["heart_disease_present"] != y_pred).sum(),
          100*(1-(X_test["heart_disease_present"] != y_pred).sum()/X_test.shape[0])
))

Number of mislabeled points out of a total 36 points : 3, performance 91.67%


As, we can see that the performance of our Classifier has increased this time to a whopping **91.67%** 

In [20]:
# Probability of Presence of Heart Disease and Probability of Absence of Heart Disease
mean_heart_disease_present=np.mean(X_train["heart_disease_present"])
mean_not_heart_disease_present=1-mean_heart_disease_present
print("Heart Disease Presence prob = {:03.2f}%, Heart Disease Absent prob = {:03.2f}%"
      .format(100*mean_heart_disease_present,100*mean_not_heart_disease_present))

Heart Disease Presence prob = 41.67%, Heart Disease Absent prob = 58.33%


## Impact of Maximum Heart Rate Achieved on Presence & Absence of Heart Disease

In [24]:
# Mean and Standard Deviation of Heart Disease for Different Values of Maximum Heart Rate

mean_max_heart_rate_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==1]["max_heart_rate_achieved"])
std_max_heart_rate_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==1]["max_heart_rate_achieved"])
mean_max_heart_rate_not_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==0]["max_heart_rate_achieved"])
std_max_heart_rate_not_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==0]["max_heart_rate_achieved"])

print("mean_max_heart_rate_heart_disease_present = {:03.2f}".format(mean_max_heart_rate_heart_disease_present))
print("std_max_heart_rate_heart_disease_present = {:03.2f}".format(std_max_heart_rate_heart_disease_present))
print("mean_max_heart_rate_not_heart_disease_present = {:03.2f}".format(mean_max_heart_rate_not_heart_disease_present))
print("std_max_heart_rate_not_heart_disease_present = {:03.2f}".format(std_max_heart_rate_not_heart_disease_present))


mean_max_heart_rate_heart_disease_present = 140.50
std_max_heart_rate_heart_disease_present = 22.74
mean_max_heart_rate_not_heart_disease_present = 157.27
std_max_heart_rate_not_heart_disease_present = 19.37


So, seems like the maximum heart rate achieved during any physical stress test seems to be on the lower side for the patient with the probability of heart disease present. On the other hand patient with a healthy heart and with lower risk of presence of heart disease has a higher maximum heart rate achieved during any physical exercise or physical stress test.

## Model's Performance with Impact of Maximum Heart Rate Achieved

In [28]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
used_features =["max_heart_rate_achieved"]
y_pred = gnb.fit(X_train[used_features].values, X_train["heart_disease_present"]).predict(X_test[used_features])
print("Number of mislabeled points out of a total {} points : {}, performance {:05.2f}%"
      .format(
          X_test.shape[0],
          (X_test["heart_disease_present"] != y_pred).sum(),
          100*(1-(X_test["heart_disease_present"] != y_pred).sum()/X_test.shape[0])
))

Number of mislabeled points out of a total 36 points : 15, performance 58.33%


## Impact of Resting Blood Pressure on Presence & Absence of Heart Disease

In [25]:
# Mean and Standard Deviation of Heart Disease for Different Values of Resting Blood Pressure

mean_resting_blood_pressure_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==1]["resting_blood_pressure"])
std_resting_blood_pressure_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==1]["resting_blood_pressure"])
mean_resting_blood_pressure_not_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==0]["resting_blood_pressure"])
std_resting_blood_pressure_not_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==0]["resting_blood_pressure"])

print("mean_resting_blood_pressure_heart_disease_present = {:03.2f}".format(mean_resting_blood_pressure_heart_disease_present))
print("std_resting_blood_pressure_heart_disease_present = {:03.2f}".format(std_resting_blood_pressure_heart_disease_present))
print("mean_resting_blood_pressure_not_heart_disease_present = {:03.2f}".format(mean_resting_blood_pressure_not_heart_disease_present))
print("std_resting_blood_pressure_not_heart_disease_present = {:03.2f}".format(std_resting_blood_pressure_not_heart_disease_present))

mean_resting_blood_pressure_heart_disease_present = 133.20
std_resting_blood_pressure_heart_disease_present = 18.03
mean_resting_blood_pressure_not_heart_disease_present = 129.96
std_resting_blood_pressure_not_heart_disease_present = 16.95


So, for obvious reason it seems like the mean of the resting BP for patient with heart disease is on the higher side than the mean of the resting blood pressure for patient with less resting blood pressure.  

## Model's Performance with Impact of Resting Blood Pressure

In [29]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
used_features =["resting_blood_pressure"]
y_pred = gnb.fit(X_train[used_features].values, X_train["heart_disease_present"]).predict(X_test[used_features])
print("Number of mislabeled points out of a total {} points : {}, performance {:05.2f}%"
      .format(
          X_test.shape[0],
          (X_test["heart_disease_present"] != y_pred).sum(),
          100*(1-(X_test["heart_disease_present"] != y_pred).sum()/X_test.shape[0])
))

Number of mislabeled points out of a total 36 points : 21, performance 41.67%


## Impact of Blood Cholesterol on Presence & Absence of Heart Disease

In [26]:
# Mean and Standard Deviation of Heart Disease for Different Values of Maximum Heart Rate

mean_cholesterol_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==1]["serum_cholesterol_mg_per_dl"])
std_cholesterol_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==1]["serum_cholesterol_mg_per_dl"])
mean_cholesterol_not_heart_disease_present = np.mean(X_train[X_train["heart_disease_present"]==0]["serum_cholesterol_mg_per_dl"])
std_cholesterol_not_heart_disease_present = np.std(X_train[X_train["heart_disease_present"]==0]["serum_cholesterol_mg_per_dl"])

print("mean_cholesterol_heart_disease_present = {:03.2f}".format(mean_cholesterol_heart_disease_present))
print("std_cholesterol_heart_disease_present = {:03.2f}".format(std_cholesterol_heart_disease_present))
print("mean_cholesterol_not_heart_disease_present = {:03.2f}".format(mean_cholesterol_not_heart_disease_present))
print("std_cholesterol_not_heart_disease_present = {:03.2f}".format(std_cholesterol_not_heart_disease_present))

mean_cholesterol_heart_disease_present = 252.43
std_cholesterol_heart_disease_present = 45.92
mean_cholesterol_not_heart_disease_present = 246.87
std_cholesterol_not_heart_disease_present = 59.92


Okay, so for obvious reason the presence of heart disease inceases with the increase in the mean of the cholesterol present in blood. 

## Model's Performance with Impact of Blood Cholesterol

In [30]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
used_features =["serum_cholesterol_mg_per_dl"]
y_pred = gnb.fit(X_train[used_features].values, X_train["heart_disease_present"]).predict(X_test[used_features])
print("Number of mislabeled points out of a total {} points : {}, performance {:05.2f}%"
      .format(
          X_test.shape[0],
          (X_test["heart_disease_present"] != y_pred).sum(),
          100*(1-(X_test["heart_disease_present"] != y_pred).sum()/X_test.shape[0])
))

Number of mislabeled points out of a total 36 points : 20, performance 44.44%


It seems that the Model's performance was at it's best when the used feature is Maximum Heart Rate Acheieved during a physical stress test for patient to be classified as Patient with Heart Disease and Patient without Heart Disease.